### PR TITLE
Start python API Solver documentation

### PR DESCRIPTION
**Pull request by alex-ozdemir**

### CODE REVIEW

**Comment by nafur** [comment]
There is another issue: https://github.com/cvc5/cvc5/pull/7064/checks?check_run_id=3536137303#step:20:34
I think the `python.rst` needs to be in `docs/api/python/`, not `docs/python/`.

File: src/api/python/cvc5.pxi [hunk 1]
```diff
@@ -506,11 +507,15 @@ cdef class Solver:
         del self.csolver
 
     def getBooleanSort(self):
+        """:return: the sort of booleans
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/api/python/cvc5.pxi]: Okay, I think that at this point everything should be consistent with the C++ documentation unless there is a reason (e.g., something python specific).

**Comment by alex-ozdemir** [comment]
Thanks for pointing that out, Gereon. I've fixed it as you suggest.

File: src/api/python/cvc5.pxi [hunk 2]
```diff
@@ -1135,17 +1406,57 @@ cdef class Solver:
...
+        :param: symbol the name of the universal variable
+        :return: the universal variable
+        """
         cdef Term term = Term(self)
         term.cterm = self.csolver.mkSygusVar(sort.csort, symbol.encode())
         return term
 
     def addSygusConstraint(self, Term t):
+        """
+        Add a forumla to the set of Sygus constraints.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1424]:
```suggestion
        Add a formula to the set of SyGuS constraints.
```

File: src/api/python/cvc5.pxi [hunk 3]
```diff
@@ -1135,17 +1406,57 @@ cdef class Solver:
...
+        SyGuS v2::
+
+            ( constraint <term> )
+
+        :param: term the formula to add as a constraint
+        """
         self.csolver.addSygusConstraint(t.cterm)
 
     def addSygusInvConstraint(self, Term inv_f, Term pre_f, Term trans_f, Term post_f):
+        """Add a set of Sygus constraints to the current state that correspond to an
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1434]:
```suggestion
        """Add a set of SyGuS constraints to the current state that correspond to an
```

File: src/api/python/cvc5.pxi [hunk 4]
```diff
@@ -1252,18 +1608,38 @@ cdef class Solver:
...
+        Define n-ary function.
         Supports two uses:
-                Term defineFun(str symbol, List[Term] bound_vars,
-                               Sort sort, Term term, bool glbl)
-                Term defineFun(Term fun, List[Term] bound_vars,
-                               Term term, bool glbl)
-        '''
+
+        - Term defineFun(str symbol, List[Term] bound_vars, Sort sort, Term term, bool glbl)
+        - Term defineFun(Term fun, List[Term] bound_vars, Term term, bool glbl)
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1630]:
```suggestion
        - ``Term defineFun(str symbol, List[Term] bound_vars, Sort sort, Term term, bool glbl)``
        - ``Term defineFun(Term fun, List[Term] bound_vars, Term term, bool glbl)``
```

**Review by nafur — COMMENTED** [review]
Thanks for all the work!

of course some requests :)
- it should be `:param name: description`, not `:param: name description`
- the format indicators (`SyGuS` and `SMT-LIB`) are not on their own line. Maybe add a newline before them?
- the docstring is not adopted by decorators (`checkEntailed`, `checkSatAssuming`). There are ways to avoid this, check https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-automodule (search for "decorator")
- It would be great if you could turn the (smtlib/sygus) code examples into code blocks (https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-code-block). We have proper highlighting for `smtlib` in place, though we probably need to add the SyGuS command (https://github.com/cvc5/cvc5/blob/master/docs/ext/smtliblexer.py#L18)

File: src/api/python/cvc5.pxi [hunk 5]
```diff
@@ -1284,13 +1704,28 @@ cdef class Solver:
...
+
         Supports two uses:
-                Term defineFunRec(str symbol, List[Term] bound_vars,
-                               Sort sort, Term term, bool glbl)
-                Term defineFunRec(Term fun, List[Term] bound_vars,
-                               Term term, bool glbl)
-        '''
+
+        - Term defineFunRec(str symbol, List[Term] bound_vars, Sort sort, Term term, bool glbl)
+        - Term defineFunRec(Term fun, List[Term] bound_vars, Term term, bool glbl)
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1712]:
```suggestion
        - ``Term defineFunRec(str symbol, List[Term] bound_vars, Sort sort, Term term, bool glbl)``
        - ``Term defineFunRec(Term fun, List[Term] bound_vars, Term term, bool glbl)``
```
or maybe instead use ```:cpp:func:`Solver::defineFunRec()` ```?

File: src/api/python/cvc5.pxi [hunk 6]
```diff
@@ -1284,13 +1704,28 @@ cdef class Solver:
...
+
+
+        SMT-LIB:
+
+        .. code-block:: smtlib
+
+            ( define-funs-rec ( <function_decl>^{n+1} ) ( <term>^{n+1} ) )
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1719]:
```suggestion
            ( define-funs-rec ( <function_decl>^n ) ( <term>^n ) )
```

File: src/api/python/cvc5.pxi [hunk 7]
```diff
@@ -1311,6 +1746,22 @@ cdef class Solver:
         return term
 
     def defineFunsRec(self, funs, bound_vars, terms):
+        """Define recursive functions.
+
+        SMT-LIB:
+
+        .. code-block:: smtlib
+
+            ( define-funs-rec ( <function_decl>^{n+1} ) ( <term>^{n+1} ) )
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1755]:
```suggestion
            ( define-funs-rec ( <function_decl>^n ) ( <term>^n ) )
```

File: src/api/python/cvc5.pxi [hunk 8]
```diff
@@ -1311,6 +1746,22 @@ cdef class Solver:
...
+
+        Create elements of parameter 'funs' with mkConst().
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1757]:
```suggestion
        Create elements of parameter ``funs`` with mkConst().
```

File: docs/ext/smtliblexer.py [hunk 9]
```diff
@@ -23,6 +23,9 @@ class SmtLibLexer(RegexLexer):
         'get-assignment', 'get-info', 'get-model', 'get-option', 'get-proof',
         'get-unsat-assumptions', 'get-unsat-core', 'get-value', 'pop', 'push',
         'reset', 'reset-assertions', 'set-info', 'set-logic', 'set-option',
+        # SyGuS v2
+        'declare-var', 'constraint', 'inv-constraint', 'synth-fun',
+        'check-synth', 'synth-inv',
```
Comment (nafur) [inline docs/ext/smtliblexer.py:28]:
```suggestion
        'declare-var', 'constraint', 'inv-constraint', 'synth-fun',
        'check-synth', 'synth-inv', 'declare-pool',
```

File: src/api/python/cvc5.pxi [hunk 10]
```diff
@@ -935,22 +1079,38 @@ cdef class Solver:
...
-        '''
-            Supports the following arguments:
-            Term mkBitVector(int size, int val=0)
-            Term mkBitVector(int size, string val, int base)
-         '''
+        """
+        Supports the following arguments:
+
+        - Term mkBitVector(int size, int val=0)
+        - Term mkBitVector(int size, string val, int base)
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1107]: Ditto.

File: src/api/python/cvc5.pxi [hunk 11]
```diff
@@ -806,14 +898,15 @@ cdef class Solver:
...
-                Op mkOp(Kind kind, const string& arg)
-                Op mkOp(Kind kind, uint32_t arg)
-                Op mkOp(Kind kind, uint32_t arg0, uint32_t arg1)
-        '''
+
+        - Op mkOp(Kind kind)
+        - Op mkOp(Kind kind, Kind k)
+        - Op mkOp(Kind kind, const string& arg)
+        - Op mkOp(Kind kind, uint32_t arg)
+        - Op mkOp(Kind kind, uint32_t arg0, uint32_t arg1)
```
Comment (nafur) [inline src/api/python/cvc5.pxi:908]: Ditto.

File: src/api/python/cvc5.pxi [hunk 12]
```diff
@@ -871,21 +986,20 @@ cdef class Solver:
...
+        :param: `den` if not None, the value is `val`/`den`
+        :return: a real term with literal value
+
+        Can be used in various forms:
+
+        - Given a string "N/D" constructs the corresponding rational.
+        - Given a string "W.D" constructs the reduction of (W * P + D)/P, where P is the appropriate power of 10.
+        - Given a float f, constructs the rational matching f's string representation. This means that mkReal(0.3) gives 3/10 and not the IEEE-754 approximation of 3/10.
+        - Given a string "W" or an integer, constructs that integer.
+        - Given two strings and/or integers N and D, constructs N/D.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1001]:
```suggestion
        - Given a string ``"N/D"`` constructs the corresponding rational.
        - Given a string ``"W.D"`` constructs the reduction of ``(W * P + D)/P``, where ``P`` is the appropriate power of 10.
        - Given a float ``f``, constructs the rational matching ``f``'s string representation. This means that ``mkReal(0.3)`` gives ``3/10`` and not the IEEE-754 approximation of ``3/10``.
        - Given a string ``"W"`` or an integer, constructs that integer.
        - Given two strings and/or integers ``N`` and ``D``, constructs ``N/D``.
```

File: src/api/python/cvc5.pxi [hunk 13]
```diff
@@ -989,46 +1149,96 @@ cdef class Solver:
...
+        :param sig: Number of bits in the significand
+        :return: the floating-point constant
+        """
         cdef Term term = Term(self)
         term.cterm = self.csolver.mkNegZero(exp, sig)
         return term
 
     def mkRoundingMode(self, RoundingMode rm):
+        """Create a roundingmode constant.
+        :param rm: the floating point rounding mode this constant represents
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1220]:
```suggestion
        """Create a roundingmode constant.
        
        :param rm: the floating point rounding mode this constant represents
```

File: src/api/python/cvc5.pxi [hunk 14]
```diff
@@ -898,31 +1012,61 @@ cdef class Solver:
...
+        :param sort: the sort of the nil term
+        :return: the separation logic nil term
+        """
         cdef Term term = Term(self)
         term.cterm = self.csolver.mkSepNil(sort.csort)
         return term
 
     def mkString(self, str s, useEscSequences = None):
+        """Create a String constant from a `str` which may contain SMT-LIB
+        compatible escape sequences like `\u1234` to encode unicode characters.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1064]:
```suggestion
        compatible escape sequences like `\\u1234` to encode unicode characters.
```

File: src/api/python/cvc5.pxi [hunk 15]
```diff
@@ -769,14 +854,15 @@ cdef class Solver:
...
-                    Term mkTerm(Kind kind, List[Term] children)
-
-                where List[Term] can also be comma-separated arguments
-        '''
+        """
+        Supports the following arguments:
+
+        - Term mkTerm(Kind kind)
+        - Term mkTerm(Kind kind, Op child1, List[Term] children)
+        - Term mkTerm(Kind kind, List[Term] children)
```
Comment (nafur) [inline src/api/python/cvc5.pxi:862]: Ditto.

File: src/api/python/cvc5.pxi [hunk 16]
```diff
@@ -989,46 +1149,96 @@ cdef class Solver:
...
+        """
         cdef Term term = Term(self)
         term.cterm = self.csolver.mkRoundingMode(<c_RoundingMode> rm.crm)
         return term
 
     def mkUninterpretedConst(self, Sort sort, int index):
+        """Create uninterpreted constant.
+
+        @param sort Sort of the constant
+        @param index Index of the constant
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1230]:
```suggestion
        :param sort: Sort of the constant
        :param index: Index of the constant
```

File: src/api/python/cvc5.pxi [hunk 17]
```diff
@@ -1397,21 +1960,83 @@ cdef class Solver:
         return term
 
     def pop(self, nscopes=1):
+        """Pop (a) level(s) from the assertion stack.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1963]:
```suggestion
        """Pop ``nscopes`` level(s) from the assertion stack.
```

File: src/api/python/cvc5.pxi [hunk 18]
```diff
@@ -1397,21 +1960,83 @@ cdef class Solver:
...
+        .. code-block:: smtlib
+
+            ( pop <numeral> )
+
+        :param nscopes: the number of levels to pop
+        """
         self.csolver.pop(nscopes)
 
     def push(self, nscopes=1):
+        """ Push (a) level(s) to the assertion stack.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1976]:
```suggestion
        """ Push ``nscopes`` level(s) to the assertion stack.
```

**Comment by alex-ozdemir** [comment]
Thank you for the comments, Gereon. I believe that I have addressed them.

File: src/api/python/cvc5.pxi [hunk 19]
```diff
@@ -1284,13 +1705,28 @@ cdef class Solver:
...
+        - ``Term defineFunRec(Term fun, List[Term] bound_vars, Term term, bool glbl)``
+
+
+        SMT-LIB:
+
+        .. code-block:: smtlib
+
+            ( define-funs-rec ( <function_decl>^n ) ( <term>^n ) )
+
+        Create elements of parameter 'funs' with mkConst().
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1722]:
```suggestion
        Create elements of parameter ``funs`` with mkConst().
```

File: src/api/python/cvc5.pxi [hunk 20]
```diff
@@ -1337,12 +1799,47 @@ cdef class Solver:
...
+        """
+        Get the set of unsat ("failed") assumptions.
+
+        SMT-LIB:
+
+        .. code-block:: smtlib
+
+            ( get-unsat-assumptions )
+
+        Requires to enable option 'produce-unsat-assumptions'.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1839]:
```suggestion
        Requires to enable option ``produce-unsat-assumptions``.
```

File: src/api/python/cvc5.pxi [hunk 21]
```diff
@@ -1351,6 +1848,18 @@ cdef class Solver:
...
     def getUnsatCore(self):
+        """Get the unsatisfiable core.
+
+        SMT-LIB:
+
+        .. code-block:: smtlib
+
+            ( get-unsat-core )
+
+        Requires to enable option 'produce-unsat-cores'.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1859]:
```suggestion
        Requires to enable option ``produce-unsat-cores``.
```

File: src/api/python/cvc5.pxi [hunk 22]
```diff
@@ -898,31 +1012,61 @@ cdef class Solver:
...
+        :param sort: the sort of the nil term
+        :return: the separation logic nil term
+        """
         cdef Term term = Term(self)
         term.cterm = self.csolver.mkSepNil(sort.csort)
         return term
 
     def mkString(self, str s, useEscSequences = None):
+        """Create a String constant from a `str` which may contain SMT-LIB
+        compatible escape sequences like `\\u1234` to encode unicode characters.
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1064]:
```suggestion
        compatible escape sequences like ``\\u1234`` to encode unicode characters.
```

File: src/api/python/cvc5.pxi [hunk 23]
```diff
@@ -1111,19 +1358,55 @@ cdef class Solver:
...
+            ( check-sat )
+
+        :return: the result of the satisfiability check.
+        """
         cdef Result r = Result()
         r.cr = self.csolver.checkSat()
         return r
 
     def mkSygusGrammar(self, boundVars, ntSymbols):
+        """Create a Sygus grammar. The first non-terminal is treated as the
```
Comment (nafur) [inline src/api/python/cvc5.pxi:1403]:
```suggestion
        """Create a SyGuS grammar. The first non-terminal is treated as the
```

File: src/api/python/cvc5.pxi [hunk 24]
```diff
@@ -769,14 +854,15 @@ cdef class Solver:
...
-                where List[Term] can also be comma-separated arguments
-        '''
+        """
+        Supports the following arguments:
+
+        - ``Term mkTerm(Kind kind)``
+        - ``Term mkTerm(Kind kind, Op child1, List[Term] children)``
+        - ``Term mkTerm(Kind kind, List[Term] children)``
+
+        where List[Term] can also be comma-separated arguments
```
Comment (nafur) [inline src/api/python/cvc5.pxi:864]:
```suggestion
        where ``List[Term]`` can also be comma-separated arguments
```

**Review by nafur — APPROVED** [review]
LGTM, thanks for making all these changes.

**Pull request merged**